### PR TITLE
WIP: Create zip archive from issue #111

### PR DESCRIPTION
**Pull request by eliroca**

### CODE REVIEW

File: src/docserv/functions.py [hunk 1]
```diff
@@ -32,6 +33,49 @@ def mail(text, subject, to):
     p.communicate(msg.as_string())
 
 
+def get_all_file_paths(directory, zip_formats):
```
Comment (eliroca) [inline src/docserv/functions.py]: This function looks weird, could't find a prettier way. Help would be welcome :)

File: src/docserv/functions.py [hunk 2]
```diff
@@ -32,6 +33,49 @@ def mail(text, subject, to):
...
+            filepath = os.path.join(root, filename)
+            format_ok = 0
+            for format in zip_formats:
+                if format in filepath:
+                    format_ok = 1
+                elif root is directory and format in filepath:
+                    root_ok = 1
+            if format_ok:
+                file_paths.append(filepath)
+    return file_paths
```
Comment (tomschr) [inline src/docserv/functions.py]: The basic algorithm looks good, but I have only one improvement. :wink: 

For things like returning a list of files etc. it's usually a good idea to use a *generator*. So you don't return the complete list, but the the current item at the time. The changes are minimal:

```python
for root, directories, files in os.walk(directory):
    for filename in files:
        filepath = os.path.join(root, filename)
        format_ok = 0
        for format in zip_formats:
            if format in filepath:
                format_ok = 1
        if format_ok:
            yield filepath
```

File: src/docserv/functions.py [hunk 3]
```diff
@@ -32,6 +33,49 @@ def mail(text, subject, to):
...
+
+    :param str directory: a directory/folder from which files are to be collected
+    :param list zip_formats: a list containing accepted documentation formats
+        example: ["pdf", "epub", "single-html"]
+    :param str zip_file:  archive name including path where to be saved
+    :return: a list with all files matching specified documentation formats
+    :rtype: list
+    """
+    # calling function to get all file paths in the directory
+    file_paths = get_all_file_paths(directory, zip_formats)
```
Comment (tomschr) [inline src/docserv/functions.py]: If you implement the generator from above, I would remove this line...

```suggestion
```

File: src/docserv/functions.py [hunk 4]
```diff
@@ -32,6 +33,49 @@ def mail(text, subject, to):
...
+    # writing files to a zipfile
+    with ZipFile(zip_file, 'w', ZIP_DEFLATED) as zip:
+        # writing each file one by one
+        for filepath in file_paths:
```
Comment (tomschr) [inline src/docserv/functions.py]: ... and add it here:

```suggestion
        for filepath in get_all_file_paths(directory, zip_formats):
```

File: src/docserv/functions.py (hunk 1 above)
Comment (tomschr) [reply src/docserv/functions.py]: What do you mean by "weird"? The name? Or the way how it is implemented?

Comment (eliroca) [reply src/docserv/functions.py]: The way it is implemented

File: src/docserv/functions.py (hunk 2 above)
Comment (eliroca) [reply src/docserv/functions.py]: Thanks! Changes made

File: src/docserv/functions.py [hunk 5]
```diff
@@ -32,6 +33,40 @@ def mail(text, subject, to):
...
+    """
+    for root, directories, files in os.walk(directory):
+        for filename in files:
+            filepath = os.path.join(root, filename)
+            for format in zip_formats:
+                if (format in root) or (root == directory and format in filepath):
+                    yield filepath
+                    break
+
+def create_zip_file(directory, zip_formats, zip_file):
```
Comment (sknorr) [inline src/docserv/functions.py]: Afaik, this function is not yet called by anything?

**Comment by eliroca** [comment]
### After discussing with @tomschr, we came up with this solution:

Docserv creates the directory structure `lang/product/docset (en-us/caasp/2)`

Which means we could just create another directory named after the build format (pdf, epub, etc.)

So that in the end we have a directory for each format, that we can easily add to the zip archive

**No need to check for file extentions...**

```
docserv_sync_yc2ski3v
└── en-us
    └── caasp
        └── 2
            ├── epub
            ├── html
            ├── pdf
            └── single-html

```

File: src/docserv/functions.py (hunk 5 above)
Comment (eliroca) [reply src/docserv/functions.py]: nope

File: src/docserv/functions.py [hunk 6]
```diff
@@ -32,6 +33,40 @@ def mail(text, subject, to):
...
+    :param str directory: a directory from which files are to be collected
+    :param list zip_formats: a list containing accepted documentation formats
+        example: ["pdf", "epub", "single-html"]
+    :return: a generator of matching file paths
+    """
+    for root, directories, files in os.walk(directory):
+        for filename in files:
+            filepath = os.path.join(root, filename)
+            for format in zip_formats:
+                if (format in root) or (root == directory and format in filepath):
```
[the remark this answers is not in the document]
Comment (eliroca) [reply src/docserv/functions.py]: I agree. It is dodgy. See my comment below :)

File: src/docserv/functions.py [hunk 7]
```diff
@@ -32,6 +33,49 @@ def mail(text, subject, to):
...
+    """Collect all files with paths in specified formats from given directory.
+
+    :param str directory: a directory from which files are to be collected
+    :param list zip_formats: a list containing accepted documentation formats
+        example: ["pdf", "epub", "single-html"]
+    :return: a list with all files matching specified documentation formats
+    :rtype: list
+    """
+    file_paths = []
+    for root, directories, files in os.walk(directory):
```
[the remark this answers is not in the document]
Comment (eliroca) [reply src/docserv/functions.py]: The issue with docserv not cleaning up files when building new should be remedied now that I've changed the directory structure and copy procedures

**Comment by tomschr** [comment]
> @tomschr Merging this, because it works and it enhances building. If you have more comments, feel free to re-review though.

Thanks. No comments at the moment. :-)

**Pull request merged**